python2/koans/about_classes.py

```diff
@@ -62,12 +62,14 @@ def get_name(self):
 
         name = property(get_name, set_name)
 
+#Comment: I'm not that sure on how property works
```

Comment:
`property(getter, setter)` is one of Python's way of implementing getters and setters. If you're not familiar with the concept of getters and setters, you can find an explanation [here](https://guide.freecodecamp.org/java/getters-and-setters/). The example is in Java but the same concepts apply.

Getters and setters allow you to have control on how class properties are accessed and manipulated and are typically used to expose private properties in a controlled way. This allows you to even apply some logic before setting an attribute. 

As an example, consider a class with a property that should **always** be a string. 

```python
In [1]: class Test(object):
   ...:   def __init__(self):
   ...:     # This value should ALWAYS contain a string
   ...:     self._string = ''
   ...: 
   ...:   def get_string(self):
   ...:     return self._string
   ...: 
   ...:   def set_string(self, value):
   ...:     """
   ...:     This setter makes sure that whatever value is set
   ...:     to `self.string` is converted to a string.
   ...:     """
   ...:     self._string = str(value)
   ...: 
   ...:   string = property(get_string, set_string)
   ...:   

In [2]: # Create an instance
   ...: t = Test()

In [3]: # Check the type and contents of string
   ...: print(t.string.__class__.__name__, t.string)
('str', '')

In [4]: # Try to set the value of string to a number
   ...: t.string = 1234

In [5]: # Check the type and contents of string
   ...: # Notice that it's a string and not a number
   ...: print(t.string.__class__.__name__, t.string)
('str', '1234')
```

Notice that we're getting and setting the value of the private property `self._string`, but we never access that property directly; all our calls were to `t.value` and not `t._value`. Getters and setters act as guards that control how private properties are accessed. 

As another example, you can create a private attribute whose setter always raises an error. This would prevent someone from directly changing the value of a property, and force them to use some other class method to update the value.

```python
In [1]: class Test(object):
   ...:   def __init__(self):
   ...:     # This value should not be accessed directly
   ...:     self._value = 0
   ...: 
   ...:   def get_value(self):
   ...:     return self._value
   ...: 
   ...:   def set_value(self, value):
   ...:     """
   ...:     The value should never be updated directly.
   ...:     """
   ...:     raise Exception('self._value cannot be accessed directly.')
   ...: 
   ...:   def add(self, amount):
   ...:     """
   ...:     Increases this instance's value by a certain amount.
   ...:     """
   ...:     self._value += amount
   ...: 
   ...:   def subtract(self, amount):
   ...:     """
   ...:     Decreases this instance's value by a certain amount.
   ...:     """
   ...:     self._value -= amount
   ...: 
   ...:   # Assign a getter and setter for self._value
   ...:   value = property(get_value, set_value)
   ...:   

In [2]: # Create an instance
   ...: t = Test()

In [3]: # Check the initial contents of value
   ...: print(t.value)
0

In [4]: # Try to assign a value directly
   ...: try:
   ...:   # This will fail
   ...:   t.value = 100
   ...: except Exception as e:
   ...:   print(str(e))
   ...:   
self._value cannot be accessed directly.

In [5]: # Increase the value by 10
   ...: t.add(10)
   ...: print(t.value)
   ...: 
10

In [6]: # Decrease the value by 5
   ...: t.subtract(5)
   ...: print(t.value)
   ...: 
5
```

Python doesn't really have the concept of private attributes. Prefixing an attribute with an underscore only suggests that it should not be accessed directly, but does not prevent someone from doing so. The guards we set in the previous example can easily be bypassed by accessing the private attribute `t._value` directly.

```python
In [1]: class Test(object):
   ...:   def __init__(self):
   ...:     # This value should not be accessed directly
   ...:     self._value = 0
   ...: 
   ...:   def get_value(self):
   ...:     return self._value
   ...: 
   ...:   def set_value(self, value):
   ...:     """
   ...:     The value should never be updated directly.
   ...:     """
   ...:     raise Exception('self._value cannot be accessed directly.')
   ...: 
   ...:   def add(self, amount):
   ...:     """
   ...:     Increases this instance's value by a certain amount.
   ...:     """
   ...:     self._value += amount
   ...: 
   ...:   def subtract(self, amount):
   ...:     """
   ...:     Decreases this instance's value by a certain amount.
   ...:     """
   ...:     self._value -= amount
   ...: 
   ...:   # Assign a getter and setter for self._value
   ...:   value = property(get_value, set_value)
   ...:   

In [2]: # Create an instance
   ...: t = Test()

In [3]: # Access the private value directly
   ...: t._value = 1000
   ...: print(t.value)
   ...: 
1000
```

Coming back to `property`, it provides a way of generating an object that acts as a getter and setter for a property. `property(<getter function>, <setter function>)` returns an object that calls the getter function when the value is accessed, and the setter function when a value is assigned to it.
```python
# Accessing a `property` object calls the getter function
`t.value` calls `t.get_value()`
# Assigning to a `property` object calls the setter function
`t.value = 100` calls `t.set_value(100)`
```

Reply:
Great explanation!